grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractElement.java

```diff
@@ -125,6 +132,47 @@ public boolean equals(Object object) {
     }
 
     /**
+     * Sets the value of a property with the added restriction that no other vertex can have that property.
+     *
+     * @param key The key of the unique property to mutate
+     * @param value The new value of the unique property
+     */
+    void propertyUnique(P key, String value){
+        if(!graph().isBatchGraph()) {
+            GraphTraversal<Vertex, Vertex> traversal = graph().getTinkerTraversal().has(key.name(), value);
+            if(traversal.hasNext()) throw PropertyNotUniqueException.cannotChangeProperty(element(), traversal.next(), key, value);
+        }
+
+        property(key, value);
+    }
+
+    /**
+     * Sets a property which cannot be mutated
+     *
+     * @param property The key of the immutable property to mutate
+     * @param newValue The new value to put on the property (if the property is not set)
+     * @param foundValue The current value of the property
+     * @param converter Helper method to ensure data is persisted in the correct format
+     */
+    <X> void propertyImmutable(P property, X newValue, X foundValue, Function<X, Object> converter){
```

Comment:
It makes it easier to handle comparing the properties using one type and persisting them using another. For example with resources of type date they are persisted as longs but compared as time stamps.